anker/src/state.rs

```diff
@@ -90,6 +90,74 @@ pub struct HistoricalStSolPrice {
     pub st_sol_price_in_ust: MicroUst,
 }
 
+#[repr(C)]
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    BorshDeserialize,
+    BorshSerialize,
+    BorshSchema,
+    Eq,
+    PartialEq,
+    Serialize,
+)]
+pub struct HistoricalStSolPriceArray(pub [HistoricalStSolPrice; POOL_PRICE_NUM_SAMPLES]);
+
+impl HistoricalStSolPriceArray {
+    /// Create new `HistorialStSolPriceArray` with slot 0 and 1 UST in each
+    /// position of the array.
+    pub fn new() -> Self {
+        HistoricalStSolPriceArray(
+            [HistoricalStSolPrice {
+                slot: 0,
+                st_sol_price_in_ust: MicroUst(1_000_000),
+            }; 5],
+        )
+    }
+
+    /// Get last price from the array.
+    pub fn last(&self) -> HistoricalStSolPrice {
+        self.0[POOL_PRICE_NUM_SAMPLES - 1]
+    }
+
+    /// Insert `st_sol_price_in_ust` at the end of the array and rotate it.
+    pub fn insert_and_rotate(&mut self, slot: Slot, st_sol_price_in_ust: MicroUst) {
+        // Maintain the invariant that samples are sorted by ascending slot number.
+        // The sample at index 0 is the oldest, so we remove it (well, move it to the
+        // end to be overwritten), and move everything else closer to the beginning
+        // of the array. Then we overwrite the last element with the current price
+        // and slot number, and we confirmed above that that slot number is larger
+        // than the slot number of the sample before it.
+        self.0.rotate_left(1);
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].slot = slot;
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].st_sol_price_in_ust = st_sol_price_in_ust;
```

Comment:
The invariant should be preserved if we pass in the current clock, but for defensive programming, let’s explicitly assert it as well:

```rust
assert!(self.0[POOL_PRICE_NUM_SAMPLES - 1].slot >= self.0[POOL_PRICE_NUM_SAMPLES - 2]);
```